### ISSUE TITLE
loss.backward(retain_graph=True) in optimization of G ?

### ISSUE THREAD

**Issue by JancisWang** [MutatedFlood/text-summary#1]
why in the optimization of G, loss.backward() should maintain the computation graph? (retain_graph=True)? What does this use for? And can you give some paper links for the technique you use in Q.py? Thank you!

**Comment by rentruewang** [comment]
@JancisWang

First of all, my apologies for not replying to you, I didn't know how to use GitHub back then. Now I feel awful.

The reason for `loss.backward(retain_graph=True)` is that: every time after you do a `loss.backward()`, the computational graph is freed afterwards, that is, PyTorch no longer knows how the end result (`loss`) is related to its sources (layer parameters etc). As a result, you can't call `loss.backward()` twice.

Observe that the `loss.backward(retain_graph=True)` happens in a loop. While the graph is created only once, due to the nature of temporal difference learning, the gradient is related to the difference on every step. That's why I have to tell PyTorch explicitly not to free the computation graph, because otherwise on second `backward` the information of how the value is calculated is no longer valid. An alternative approach without using `retain_graph=True` would be to sum all the losses and calculate the gradient at once. I've tried, but at the time the server I ran the code on could not handle it, and it yields slightly less readable code. Though to what extent is really subjective.

I hope this helps anyone who sees this.
